api/include/opentelemetry/trace/global_provider.h

```diff
@@ -0,0 +1,68 @@
+#pragma once
+
+#include <mutex>
+
+#include "opentelemetry/trace/noop.h"
+#include "opentelemetry/trace/tracer_provider.h"
+#include "opentelemetry/nostd/unique_ptr.h"
+
+using opentelemetry::trace::NoopTracer;
+using opentelemetry::trace::TracerProvider;
+
+std::mutex provider_mutex;
```

Comment:
Doesn't this lead to a linker error "multiple definition of `provider_mutex'" since it creates a provider_mutex in every translation unit that includes this header?

Reply:
One possible resolution: move this to `class Provider` as:

```
  static std::mutex* GetMutex() {
    static std::mutex* provider_mutex = new std::mutex;
    return provider_mutex;
  }
```

But this still depends on std::mutex.

Reply:
How about this:

```c++
class Provider
{
public:
  static TracerProvider *GetTracerProvider() { return Ptr()->load(std::memory_order_acquire); }

  static void SetTracerProvider(TracerProvider *tp) { Ptr()->store(tp, std::memory_order_release); }

private:
  static std::atomic<TracerProvider *> *Ptr()
  {
    static std::atomic<TracerProvider *> *singleton =
        new std::atomic<TracerProvider *>(new DefaultTracerProvider);
    return singleton;
  }
};
```

I think it should be okay to use `std::atomic`, IMO it doesn't introduce ABI issues.

Reply:
This seems to cause some issues when overwriting during a get - specifically it appears there's a memory leak.

I was able to get around it like so:
```
static void SetTracerProvider(TracerProvider *tp)
{
  if (auto old = Ptr()->load(std::memory_order_acquire))
  {
    delete old;
  }
  Ptr()->store(tp, std::memory_order_release);
}
```

Though this causes a couple issues that I can see:
1) When doing a set before a get it'll wastefully construct a default provider, which isn't a _huge_ deal
2) There're still some ordering issues, where a set could deconstruct the reference returned by a parallel get.

Reply:
Yeah, that's race prone. Let's try `Ptr()->exchange(...)`